subt/main.py

```diff
@@ -469,6 +468,39 @@ def on_pose2d(self, timestamp, data):
         self.xyz = x, y, z
         self.trace.update_trace(self.xyz)
 
+    def on_pose3d(self, timestamp, data):
+        xyz, rot = data
+        ypr = quaternion.euler_zyx(rot)
+        if self.offset is None:
+            # we cannot align global coordinates if offset is not known
+            return
+        x0, y0, z0 = self.offset
+        # pretend that received coordinates are in robot frame (compatible with on_pose2d)
+        xyz = xyz[0] - x0, xyz[1] - y0, xyz[2] - z0
+
+        pose = (xyz[0], xyz[1], ypr[0])
+        if self.last_position is not None:
+            self.is_moving = (self.last_position != pose)
+            dist = math.hypot(pose[0] - self.last_position[0], pose[1] - self.last_position[1])
+            direction = ((pose[0] - self.last_position[0]) * math.cos(self.last_position[2]) +
+                         (pose[1] - self.last_position[1]) * math.sin(self.last_position[2]))
+            if direction < 0:
+                dist = -dist
+        else:
+            dist = 0.0
+        self.last_position = pose
+        self.traveled_dist += dist
+        x, y, z = xyz
+        self.last_send_time = self.bus.publish('pose2d', [round((x + x0) * 1000), round((y + y0) * 1000),
```

Comment:
Is there a specific reason why we publish `pose2d`? I cannot recollect any users of this value. Or does this also belong to the "refactor pose3d later" intention?

Reply:
I still use it in lidarview ... and yes, once input `pose2d` is gone then also output `pose2d` should go